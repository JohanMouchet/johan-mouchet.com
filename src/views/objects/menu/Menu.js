// @flow

import * as React from "react";
import cx from "classnames";
import "./Menu.scss";
import { Button } from "views/objects";

type MenuItem = {
  text: string,
  url?: string,
  button?: Button,
  className?: string | Array<string> | Object,
  menu?: [MenuItem],
};

type Props = {
  menu: [MenuItem],
};

const Menu = ({ menu }: Props) => {
  const renderMenu = (menu, isSubmenu) => {
    const classNames = cx(isSubmenu ? "o-menu__submenu" : "o-menu");

    return !menu ? null : (
      <ul className={classNames}>
        {menu.map((item) => (
          <li
            className={cx(
              "o-menu__item",
              item.menu && "o-menu__item--has-children",
              item.className
            )}
            key={item.text || item.button.children}
          >
            {(item.text && (
              <a
                className={cx(
                  "o-menu__link",
                  item.menu && "o-menu__submenu-heading"
                )}
                href={item.url}
                tabIndex={!item.url ? "0" : undefined}
              >
                {item.text}
              </a>
            )) ||
              (item.button && <Button {...item.button} />) ||
              (item.menu && renderMenu(menu, true))}
            {/* TODO: Pass children to Button */}
          </li>
        ))}
      </ul>
    );
  };

  return renderMenu(menu);
};

export default Menu;
