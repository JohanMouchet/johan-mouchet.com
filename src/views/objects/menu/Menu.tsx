import * as React from "react";
import cx from "classnames";
import "./Menu.scss";
import { IconChevronDown } from "views/objects";

type Submenu = {
  text: string;
  menu?: Menu;
};

type MenuItem = {
  text: string;
  url?: string;
};

type Menu = (MenuItem & Submenu)[];

type Props = {
  menu: Menu;
  className?: string | string[] | { [key: string]: boolean };
};

const Menu: React.FC<Props> = ({ menu, className }) => {
  const renderMenu = (
    menu: Menu,
    isSubmenu?: boolean,
    className?: string | string[] | { [key: string]: boolean }
  ) => {
    const classNames = cx(isSubmenu ? "o-menu__submenu" : "o-menu", className);

    return !menu?.length ? null : (
      <ul className={classNames}>
        {menu.map((item) => (
          <li
            className={cx(
              "o-menu__item",
              item.menu && "o-menu__item--has-children"
            )}
            key={item.text}
          >
            <a
              className={cx(
                "o-menu__link",
                item.menu && "o-menu__submenu-heading"
              )}
              href={item.url}
              tabIndex={!item.url ? 0 : undefined}
            >
              {item.text}
              {item.menu && <IconChevronDown />}
            </a>

            {item.menu && renderMenu(item.menu, true)}
          </li>
        ))}
      </ul>
    );
  };

  return renderMenu(menu, false, className);
};

export default Menu;
