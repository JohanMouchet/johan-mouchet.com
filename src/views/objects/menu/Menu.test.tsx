import { render } from "@testing-library/react";
import { Menu } from "./Menu";

it("Renders <Menu /> without crashing", () => {
  expect(
    render(
      <Menu
        menu={[
          {
            text: "Menu",
            menu: [
              {
                text: "Submenu",
                url: "#",
              },
            ],
          },
        ]}
      />
    ).container.firstChild
  ).toMatchSnapshot();
});
