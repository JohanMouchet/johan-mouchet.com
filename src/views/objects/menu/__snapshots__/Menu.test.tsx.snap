// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders <Menu /> without crashing 1`] = `
<ul
  class="o-menu"
>
  <li
    class="o-menu__item o-menu__item--has-children"
  >
    <a
      class="o-menu__link o-menu__submenu-heading"
      tabindex="0"
    >
      Menu
      <svg />
    </a>
    <ul
      class="o-menu__submenu"
    >
      <li
        class="o-menu__item"
      >
        <a
          class="o-menu__link"
          href="#"
        >
          Submenu
        </a>
      </li>
    </ul>
  </li>
</ul>
`;
